reference/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,331 @@
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+
+#include "core/reorder/rcm_kernels.hpp"
+
+
+#include <algorithm>
+#include <iterator>
+#include <memory>
+#include <queue>
+#include <utility>
+#include <vector>
+
+
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/permutation.hpp>
+#include <ginkgo/core/matrix/sparsity_csr.hpp>
+
+
+namespace gko {
+namespace kernels {
+namespace reference {
+/**
+ * @brief The reordering namespace.
+ *
+ * @ingroup reorder
+ */
+namespace rcm {
+
+
+template <typename ValueType, typename IndexType>
+void get_degree_of_nodes(
+    std::shared_ptr<const ReferenceExecutor> exec,
+    std::shared_ptr<matrix::SparsityCsr<ValueType, IndexType>> adjacency_matrix,
+    std::shared_ptr<gko::Array<IndexType>> node_degrees)
+{
+    auto num_rows = adjacency_matrix->get_size()[0];
+    auto adj_ptrs = adjacency_matrix->get_row_ptrs();
+    auto node_deg = node_degrees->get_data();
+
+    for (auto i = 0; i < num_rows; ++i) {
+        node_deg[i] = adj_ptrs[i + 1] - adj_ptrs[i];
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_RCM_GET_DEGREE_OF_NODES_KERNEL);
+
+
+/**
+ * Computes a level structure rooted at `root`, returning the height of the
+ * structure along with the index of the node of minimal degree in its last
+ * level.
+ */
+template <typename IndexType, typename ValueType>
+std::pair<IndexType, size_type> rls_contender_and_height(
+    std::shared_ptr<const ReferenceExecutor> exec, IndexType root,
+    std::shared_ptr<matrix::SparsityCsr<ValueType, IndexType>> adjacency_matrix,
+    std::shared_ptr<Array<IndexType>> node_degrees, size_type num_vertices)
+{
+    auto adj_ptrs = adjacency_matrix->get_row_ptrs();
+    auto adj_idxs = adjacency_matrix->get_col_idxs();
+    auto node_degrees_p = node_degrees->get_data();
+
+    // This could actually be allocated in the calling scope, then reused.
+    auto visited_local =
+        std::unique_ptr<Array<bool>>(new Array<bool>(exec, num_vertices));
```

Comment:
```suggestion
    auto visited_local = Array<bool>(exec, num_vertices);
```
We don't need a pointer here, since it never leaves the scope.

Reply:
Agreed.